ユーザー管理機能

unitoro0903 opened the pull request:
最終課題ユーザー管理機能の実装
https://gyazo.com/884199f5b184c0ce4d06bd21294daacd
https://gyazo.com/c9335b5da8f9032fd33e6cea73da76ee
https://gyazo.com/8d3b7eec46d3f7f68c9098c64c400782
https://gyazo.com/301d1b1b4a4299828f1922b837acba8f
https://gyazo.com/640ba24be1d7cdb647ac76cb83428ca1
https://gyazo.com/8208d6d72564063eecbfb7c4a9b6140e

unitoro0903 committed:
viewファイル移動

unitoro0903 committed:
デバイスのダウンロードと新規登録画面viewファイルダウンロード

unitoro0903 committed:
ストロングパラメーター

unitoro0903 committed:
新規登録画面のHTMLのURLの設定

unitoro0903 committed:
対応するカラムの設定とモデルにバリデーション

unitoro0903 committed:
itemsのindexのビューファイルのルーティング、コントローラー設定

unitoro0903 committed:
画面遷移のパスの修正

unitoro0903 committed:
ユーザー管理機能テスト

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます！
キャプチャのご提示もお礼申し上げます。

**実装要件の補足案内に則り、各キャプチャに何の挙動を撮影したかわかるようにタイトルをつけていただけますようお願いいたします！**

また、以下のキャプチャはエラーハンドリングに関するキャプチャかとお見受けします。
https://gyazo.com/c9335b5da8f9032fd33e6cea73da76ee
**こちら、見本アプリ同様に値に不備がある場合に赤いエラーメッセージが表示されるよう実装を見直しましょう！**


実装、挙動をご確認の上、再度Gyazoで撮影いただいたキャプチャのURLをこちらのプルリクエストに添付いただければと存じます。
その後、再度レビュー依頼いただけますと幸いです。

何卒よろしくお願いいたします。

unitoro0903 commented:
- [https://gyazo.com/8d3b7eec46d3f7f68c9098c64c400782 ] 必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
- [ https://gyazo.com/d554dc6e71877f8fb2e8660630ef899d] 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
- [https://gyazo.com/640ba24be1d7cdb647ac76cb83428ca1 ] 必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
- [https://gyazo.com/8208d6d72564063eecbfb7c4a9b6140e ] 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
- [https://gyazo.com/301d1b1b4a4299828f1922b837acba8f ] ログアウトに成功する動画
- [ https://gyazo.com/884199f5b184c0ce4d06bd21294daacd] テストに成功する画像
ご確認お願いいたします。

unitoro0903 committed:
新規登録エラーハンドリング実装

minachanchan commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+         validates :first_name_kanji, presence: true
+         validates :family_name_kana, presence: true
+         validates :first_name_kana, presence: true
+         validates :birthday, presence: true
```
現在必要なバリデーションの記述が不足しております。
実装要件に則り、必要なバリデーションについて追記しましょう！

minachanchan commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+         validates :birthday, presence: true
+         
+         has_many :items
+         has_many :records
```
```ruby
         has_many :items
         has_many :records
```
こちらは該当機能実装時に差分として出力するため、今は一度コメントアウトしましょう！

minachanchan commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+<%= render "shared/second-header"%>
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
```erb
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```

こちらのコメントアウトは実装案内なので削除しておきましょう。
完成したアプリを公開する際に実装案内の記述が残っているのは望ましくないためです。
実装を終え次第、削除するようにしていきましょう！
**こちら以外にも別のファイルの記述含め、「〜しましょう」のような実装案内文が残っているようですので全て削除しましょう！**

minachanchan commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,92 @@
+          @user.email = 'testmail'
+          @user.valid?
+          expect(@user.errors.full_messages).to include('Email is invalid')
+      end
```
現在テスト項目が不足しております。
要件に沿って以下の項目全ての異常系テストを実装できているかを確認してください。
```
 ・nicknameが空では登録できない
 ・メールアドレスが空では登録できない
 ・重複したメールアドレスは登録できない
 ・メールアドレスに@を含まない場合は登録できない
 ・パスワードが空では登録できない
 ・パスワードが6文字未満では登録できない
 ・英字のみのパスワードでは登録できない
 ・数字のみのパスワードでは登録できない
 ・全角文字を含むパスワードでは登録できない
 ・パスワードとパスワード（確認用）が不一致だと登録できない
 ・姓（全角）が空だと登録できない
 ・姓（全角）に半角文字が含まれていると登録できない
 ・名（全角）が空だと登録できない
 ・名（全角）に半角文字が含まれていると登録できない
 ・姓（カナ）が空だと登録できない
 ・姓（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
 ・名（カナ）が空だと登録できない
 ・名（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
 ・生年月日が空だと登録できない
```

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
キャプチャのご提示もお礼申し上げます！

コメントいたしましたので、ご確認頂けますと幸いでございます。

実装完了まで引き続き、一緒に頑張りましょう！
修正の後、再度レビューのご依頼頂けますようお願い申し上げます。

unitoro0903 committed:
修正1

- [https://gyazo.com/5f735dc4ae7798888c64b1962d483e20] 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
-  [https://gyazo.com/6393f0653de867026323323f8fe6ada9] テストに成功する画像

div-reviewer commented on app/models/user.rb:
```diff
@@ -0,0 +1,27 @@
+        #with_options presence: true, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/, message: '全角文字を使用してください' } do
+          #validates :family_name_kana
+          #validates :first_name_kana
+       # end
```
こちらのコメントアウトですが、不要であれば削除しましょう。

**理由**
企業様に提出する際に実装案内のコメントが残っていることは好ましくないためです。

今後の学習用に残す必要がある場合は、「学習用」、「勉強用」等のコメントを追記し、第三者にも残している理由が伝わるように工夫してみましょう。

div-reviewer commented on db/migrate/20240717110526_create_items.rb:
```diff
@@ -0,0 +1,17 @@
+class CreateItems < ActiveRecord::Migration[7.0]
```
・app/models/item.rb
・db/migrate/20240717110526_create_items.rb

こちらは商品出品で作成するモデルのため、メモ帳などに転記していただき、VScodeからは削除しましょう。
差分が上がるようにするためです。

unitoro0903 replied:
こちら最初のデプロイの際に （ルートパスへアクセスすると、ダウンロードしたview素材のapp/views/items/index.html.erbが表示されること。）この実装条件ができていないとのことで指示があって実装したのですが、方法が間違っていたのでしょうか。

div-reviewer commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname {Faker::Name.last_name}
+    email {Faker::Internet.email}
+    password {Faker::Internet.password(min_length: 6)}
```
```ruby
	    password              { Faker::Internet.password(min_length: 6) }
```

Factorybotのパスワードの値を修正しましょう。

**理由**
Fakerだけの場合、稀に「数字のみ」や「英字のみ」のパスワードが生成されてしまい、テストが失敗することが考えられます。

要件書に「パスワードは、半角英数字混合での入力が必須であること」とありますので、パスワードが必ず英数混合になるように記述をしてみましょう。

◎方法1
英数字混合のpasswordを想定した上でFakerを使用する場合は、例えば

```ruby
    password              { '1a' + Faker::Internet.password(min_length: 6) }
```
のように、確実にpasswordに英数字が含まれるように指定することが可能です。

◎方法2
Fakerを用いずに{"aaaaa1"}と直接値を入れて実装することも可能です。

■直接値を入れているやり方
```ruby
コース：応用
タイトル：RSpec〜効率的にテストコードを書こう〜

FactoryBotを使用してインスタンス生成を共通化しよう
```

div-reviewer commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,238 @@
+  describe 'ユーザー新規登録' do
+    context '新規登録できる場合' do
+      it "nicknameとemailとpasswordとpassword_confirmationとfamily_name_kanjiとfirst_name_kanjiとfamily_name_kanaとfirst_name_kanaとbirthdayが存在する" do
+        @user.family_name_kanji = "位"
+        @user.first_name_kanji = "位"
+        @user.family_name_kana = "ア"
+        @user.first_name_kana = "ア"
```
```ruby
        @user.family_name_kanji = "位"
        @user.first_name_kanji = "位"
        @user.family_name_kana = "ア"
        @user.first_name_kana = "ア"
```

spec/factories/users.rbの中に正常なデータがいるはずなので、上記を削除しても動くかと思います。
もし動かない場合は、spec/factories/users.rbの名前の値に適切なデータが入っているか確認しましょう。

unitoro0903 replied:
fakerで全角漢字だけなどの設定をしていないのでこちらを削除するとえらーになってしまいそうです。

yuri240 replied:
ご返信ありがとうございます。

こちらのエラーを解消するためには、
FactoryBotの`family_name_kanji`,`first_name_kanji` に直接値を代入していただくと良さそうです。

```ruby
    family_name_kanji { "山田" }
    first_name_kanji { "太郎" }
```

理由：
```
FactoryBotでfakerを使用することは必須では無いためです。
もしfakerを使用したい場合は、全角漢字のみを生成するような設定を行う必要がございますが、カリキュラム内では案内を行っていないため、直接値を代入していただけると良さそうです。
```

div-reviewer commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,238 @@
+      end
+
+        it '重複したemailが存在する場合は登録できない' do
+          @user.family_name_kanji = "位"
+         @user.first_name_kanji = "位"
+         @user.family_name_kana = "ア"
+         @user.first_name_kana = "ア"
```
```ruby
          @user.family_name_kanji = "位"
         @user.first_name_kanji = "位"
         @user.family_name_kana = "ア"
         @user.first_name_kana = "ア"
```

emailのテストに関係ない為削除しましょう。

**■補足**
以下のテストも同様に不要な記述は削除しましょう。

・it '英字のみのパスワードでは登録できない' do
・ it '数字のみのパスワードでは登録できない' do
・  it '全角文字を含むパスワードでは登録できない' do

unitoro0903 replied:
fakerで全角漢字だけなどの設定をしていないのでこちらを削除するとえらーになってしまいそうです。

div-reviewer commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,238 @@
+            expect(@user.errors.full_messages).to include('Password is invalid. Include both letters and numbers', 'Password must not include full-width characters.')
+          end
+
+            it '姓（全角）に半角文字が含まれていると登録できない' do
+            
+              @user.first_name_kanji = "亜"
+              @user.family_name_kana = "ア"
+              @user.first_name_kana = "ア"
+              @user.family_name_kanji = 'hankaku'
+              @user.valid?
+              expect(@user.errors.full_messages).to include('Family name kanji 全角文字を使用してください')
+            end
```
```ruby
            it '姓（全角）に半角文字が含まれていると登録できない' do

              @user.first_name_kanji = "亜"
              @user.family_name_kana = "ア"
              @user.first_name_kana = "ア"
              @user.family_name_kanji = 'hankaku'
```

姓（全角）に半角文字を入れるだけにしましょう。

**■補足**
他のテストでも不要な記述は削除しましょう。

unitoro0903 replied:
fakerで全角漢字だけなどの設定をしていないのでこちらを削除するとえらーになってしまいそうです。

div-reviewer reviewed (commented):
unitoro0903様、学習お疲れ様です。
ユーザー管理機能のレビュー依頼ありがとうございます。プルリクエストの内容を確認させていただきました。

段々とコードが良くなってきていますね。素晴らしいです。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

**ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！**
どうぞよろしくおねがいします。

unitoro0903 committed:
修正2

レビューいただきありがとうございます。
4点ほどわからない部分があり、先ほどのレビューに返信させていただいております。
特にitemsのモデルを削除してしまうと最初のホームページが消えてしまうと思っています。

yuri240 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,238 @@
+          @user.family_name_kanji = "位"
+         @user.first_name_kanji = "位"
+         @user.family_name_kana = "ア"
+         @user.first_name_kana = "ア"
```
前回のレビュアーからも指摘があったかと存じますが、
83行目以降の`family_name_kanji`や`first_name_kana`の記述は不要となりますので、
削除していただけますと良さそうです。

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
キャプチャ画像と修正点確認させていただきました。

追加でコメントさせていただきましたので、ご確認お願いします。

ご不明点や疑問点がありましたら、チャット・コネクトメンターにご相談ください。

```ruby
◎ 次回提出時のお願い ◎
お手数ではありますがテストコードを追加した後は、
改めてテストコード実行後のキャプチャ画像を撮影いただき、
こちらのプルリクエストへ添付していただくようお願いいたします。
```

unitoro0903 committed:
修正3

- [https://gyazo.com/ff19549e2a6784706a2229fb6094aed4 ] テストに成功する画像
 [https://gyazo.com/8244a59c49f1c94e31dbc5190cfab59f ] 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画

div-reviewer reviewed (commented):
頂いたレビュー依頼ですが、修正点の差分があがっていないようです。
リモートのGithubの反映が遅いだけであれば申し訳ございません。

お手数ですが`コミットとプッシュのし忘れがないか確認`して頂けますでしょうか。
修正点の差分を出して`再度googleフォームのご提出`をお願いいたします。

ご不明点ございましたら、コネクトメンターにご相談ください。
お手数ですが、どうぞよろしくおねがいします。

yuri240 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,191 @@
+            @user.password = 'ｒuby123'
+          @user.password_confirmation = 'ｒuby123'
+            @user.valid?
+            expect(@user.errors.full_messages).to include('Password is invalid. Include both letters and numbers', 'Password must not include full-width characters.')
```
114行目のエラーメッセージについて

```ruby
'Password is invalid. Include both letters and numbers'
'Password must not include full-width characters.'
```

2つのメッセージが記述されておりますので、
適切なもののみ記述していただけると良さそうです。

yuri240 reviewed (commented):
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

また、１点コメントを残させていただきましたので
確認の上でMergeしていただくようお願いいたします。

ユーザー管理機能の実装、お疲れ様でした。
引き続き、応援しております。
